### PR TITLE
Refactor the terrain and reality mesh display to use a single optimized rendering technique.  

### PR DESCRIPTION
**Pull request by RBBentley**
Reality meshes are now displayed directly from GLTF without conversion to the vertex lookup table that is used for BIM geometry.  This greatly reduces the CPU processing for reality model tile loading.   Both reality models and terrain data now use the same shader technique.  The terrain mesh creation was also optimized to avoid unnecessary object creation.   The overall result is that less "janking" when tiles are loaded and the view is being updated dynamically.

### CODE REVIEW

File: common/api/ui-framework.api.md [hunk 1]
```diff
@@ -4317,7 +4317,7 @@ export interface ProjectServices {
 }
 
 // @public @deprecated
-export const PromptField: import("react-redux").ConnectedComponent<typeof PromptFieldComponent, Pick<React.ClassAttributes<PromptFieldComponent> & PromptFieldProps, "ref" | "style" | "key" | "className" | "isInFooterMode" | "openWidget" | "onOpenWidget">>;
+export const PromptField: import("react-redux").ConnectedComponent<typeof PromptFieldComponent, Pick<React.ClassAttributes<PromptFieldComponent> & PromptFieldProps, "style" | "className" | "ref" | "isInFooterMode" | "key" | "onOpenWidget" | "openWidget">>;
```
Comment (RBBentley) [inline common/api/ui-framework.api.md:4321]: @calebmshafer - I have a set of these superfluous (changed order) in api files - are these a concern or just let them go?

Comment (DanEastBentley) [reply common/api/ui-framework.api.md:4321]: In directories with this problem (ui-framework and occasionally others):
npm run clean
npm run build
npm run extract-api

Comment (RBBentley) [reply common/api/ui-framework.api.md:4321]: Thanks Dan -- I just tried that and it seems to produce the same output... ;(

Comment (calebmshafer) [reply common/api/ui-framework.api.md:4321]: @RBBentley what node version are you running?

Comment (RBBentley) [reply common/api/ui-framework.api.md:4321]: v12.16.2

Comment (scsewall) [reply common/api/ui-framework.api.md:4321]: @RBBentley This happens to me sometimes even though I don't change code in the ui directories. I just revert the changes locally. If you push them I believe that you will get a CI build failure.

File: core/frontend/src/render/RenderSystem.ts [hunk 2]
```diff
@@ -14,6 +14,7 @@ import { SkyBox } from "../DisplayStyleState";
 import { imageElementFromImageSource } from "../ImageUtil";
 import { IModelApp } from "../IModelApp";
 import { IModelConnection } from "../IModelConnection";
+import { ToolAdmin } from "../imodeljs-frontend";
```
Comment (pmconne) [inline core/frontend/src/render/RenderSystem.ts:17]: Did your editor change this? Don't import from imodeljs-frontend inside imodeljs-frontend - import from ./tools/ToolAdmin.

File: core/frontend/src/render/RenderSystem.ts [hunk 3]
```diff
@@ -30,7 +32,7 @@ import { RenderGraphic, RenderGraphicOwner } from "./RenderGraphic";
 import { RenderMemory } from "./RenderMemory";
 import { RenderTarget } from "./RenderTarget";
 import { ScreenSpaceEffectBuilder, ScreenSpaceEffectBuilderParams } from "./ScreenSpaceEffectBuilder";
-import { ToolAdmin } from "../tools/ToolAdmin";
+import { RealityMeshGeometry } from "./webgl/RealityMesh";
```
Comment (pmconne) [inline core/frontend/src/render/RenderSystem.ts:35]: Nothing in the webgl directory should escape. Make an opaque type that exposes just the API you need, like you had with RenderTerrainMeshGeometry.

File: common/api/ui-framework.api.md (hunk 1 above)
Comment (RBBentley) [reply common/api/ui-framework.api.md:4321]: Yes - I generally revert them as well... These slipped in and I guess I'll just back them out, but it sure is annoying.

File: core/frontend/src/render/primitives/mesh/RealityMeshPrimitive.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+/** @packageDocumentation
+ * @module Rendering
+ */
+
+import { QParams2d, QParams3d, RenderTexture } from "@bentley/imodeljs-common";
+import { GltfMeshData } from "../../../tile/internal";
+import { RenderMemory } from "../../RenderMemory";
+import { Mesh } from "./MeshPrimitives";
+
+export interface RealityMeshProps {
```
Comment (pmconne) [inline core/frontend/src/render/primitives/mesh/RealityMeshPrimitive.ts:14]: What is the distinction between the classes in this file and the ones in RealityMeshPrimitive.ts?

**Review by pmconne — CHANGES_REQUESTED** [review]

File: core/frontend/src/render/webgl/glsl/FeatureSymbology.ts [hunk 5]
```diff
@@ -125,9 +129,11 @@ float computeLineCode() {
 
 function addFeatureIndex(vert: VertexShaderBuilder): void {
   vert.addGlobal("g_featureIndex", VariableType.Vec3);
-  vert.addFunction(getFeatureIndex(vert.usesInstancedGeometry));
 
-  if (!vert.usesInstancedGeometry)
+  vert.addFunction(decodeUint24);
+  vert.addFunction(getFeatureIndex(vert));
+
+  if (vert.usesVertexTable)
```
Comment (pmconne) [inline core/frontend/src/render/webgl/glsl/FeatureSymbology.ts:136]: This used to only execute for uninstanced geometry - was it intentionally changed to execute for instanced geometry too?

**Review by pmconne — CHANGES_REQUESTED** [review]
The following imodeljs-frontend test fails.
```
  1) Early Z driver bug workaround
       applies to shaders lacking discard statements when buggy driver is detected:

      AssertionError: expected false to be true
      + expected - actual

      -false
      +true

      at D:\js\s\imodeljs\core\frontend\src\test\render\webgl\EarlyZBugWorkaround.test.ts:63:54
      at RealityMeshTechnique.forEachProgram (D:\js\s\imodeljs\core\frontend\src\render\webgl\Technique.ts:250:7)
      at Techniques.forEachVariedProgram (D:\js\s\imodeljs\core\frontend\src\render\webgl\Technique.ts:894:19)
      at Context.<anonymous> (D:\js\s\imodeljs\core\frontend\src\test\render\webgl\EarlyZBugWorkaround.test.ts:62:36)
```

**Review by pmconne — CHANGES_REQUESTED** [review]
core-full-stack-tests failure:
```
  1) Background map
       obscures model based on settings:

      AssertionError: expected 1 to equal 2
      + expected - actual

      -1
      +2

      at expectPixel (D:/js/s/imodeljs/full-stack-tests/core/src/frontend/standalone/BackgroundMap.test.ts:136:37)
      at expectPixelTypes (D:/js/s/imodeljs/full-stack-tests/core/src/frontend/standalone/BackgroundMap.test.ts:142:7)
      at D:/js/s/imodeljs/full-stack-tests/core/src/frontend/standalone/BackgroundMap.test.ts:172:9
      at Object.testOnScreenViewport (D:/js/s/imodeljs/full-stack-tests/core/src/frontend/TestViewport.ts:336:5)
      at Context.<anonymous> (D:/js/s/imodeljs/full-stack-tests/core/src/frontend/standalone/BackgroundMap.test.ts:165:5)
```

File: core/frontend/src/render/RenderSystem.ts (hunk 2 above)
Comment (RBBentley) [reply core/frontend/src/render/RenderSystem.ts]: Yes... Studio Code has option to do this that I must have clicked.  Fixed.

**Review by DanEastBentley — APPROVED** [review]

File: core/frontend/src/render/RenderSystem.ts (hunk 3 above)
Comment (RBBentley) [reply core/frontend/src/render/RenderSystem.ts]: Fixed.

File: core/frontend/src/render/webgl/glsl/FeatureSymbology.ts (hunk 5 above)
Comment (RBBentley) [reply core/frontend/src/render/webgl/glsl/FeatureSymbology.ts]: Definitely an error.  Fixed.

**Comment by RBBentley** [comment]
> The following imodeljs-frontend test fails.
> 
> ```
>   1) Early Z driver bug workaround
>        applies to shaders lacking discard statements when buggy driver is detected:
> 
>       AssertionError: expected false to be true
>       + expected - actual
> 
>       -false
>       +true
> 
>       at D:\js\s\imodeljs\core\frontend\src\test\render\webgl\EarlyZBugWorkaround.test.ts:63:54
>       at RealityMeshTechnique.forEachProgram (D:\js\s\imodeljs\core\frontend\src\render\webgl\Technique.ts:250:7)
>       at Techniques.forEachVariedProgram (D:\js\s\imodeljs\core\frontend\src\render\webgl\Technique.ts:894:19)
>       at Context.<anonymous> (D:\js\s\imodeljs\core\frontend\src\test\render\webgl\EarlyZBugWorkaround.test.ts:62:36)
> ```

Fixed.

**Comment by RBBentley** [comment]
> core-full-stack-tests failure:
> 
> ```
>   1) Background map
>        obscures model based on settings:
> 
>       AssertionError: expected 1 to equal 2
>       + expected - actual
> 
>       -1
>       +2
> 
>       at expectPixel (D:/js/s/imodeljs/full-stack-tests/core/src/frontend/standalone/BackgroundMap.test.ts:136:37)
>       at expectPixelTypes (D:/js/s/imodeljs/full-stack-tests/core/src/frontend/standalone/BackgroundMap.test.ts:142:7)
>       at D:/js/s/imodeljs/full-stack-tests/core/src/frontend/standalone/BackgroundMap.test.ts:172:9
>       at Object.testOnScreenViewport (D:/js/s/imodeljs/full-stack-tests/core/src/frontend/TestViewport.ts:336:5)
>       at Context.<anonymous> (D:/js/s/imodeljs/full-stack-tests/core/src/frontend/standalone/BackgroundMap.test.ts:165:5)
> ```

Fixed.

**Comment by RBBentley** [comment]
> The following imodeljs-frontend test fails.
> 
> ```
>   1) Early Z driver bug workaround
>        applies to shaders lacking discard statements when buggy driver is detected:
> 
>       AssertionError: expected false to be true
>       + expected - actual
> 
>       -false
>       +true
> 
>       at D:\js\s\imodeljs\core\frontend\src\test\render\webgl\EarlyZBugWorkaround.test.ts:63:54
>       at RealityMeshTechnique.forEachProgram (D:\js\s\imodeljs\core\frontend\src\render\webgl\Technique.ts:250:7)
>       at Techniques.forEachVariedProgram (D:\js\s\imodeljs\core\frontend\src\render\webgl\Technique.ts:894:19)
>       at Context.<anonymous> (D:\js\s\imodeljs\core\frontend\src\test\render\webgl\EarlyZBugWorkaround.test.ts:62:36)
> ```

Fixed.

File: full-stack-tests/core/src/frontend/standalone/BackgroundMap.test.ts [hunk 6]
```diff
@@ -134,7 +134,7 @@ describe("Background map", () => {
         }
 
         expect(actualPixel.type).to.equal(Pixel.GeometryType.Surface);
-        expect(actualPixel.featureTable).not.to.be.undefined;
+        //  With the unified reality/terrain mesh, a feature table is present, so this test removed.... expect(actualPixel.featureTable).not.to.be.undefined;
```
Comment (pmconne) [inline full-stack-tests/core/src/frontend/standalone/BackgroundMap.test.ts:137]: The line you commented out was testing that the feature table **was** defined. Your comment claims that a feature table is present, so why would it need to be commented out? If the behavior has changed, please update the test to verify the new behavior.

**Review by pmconne — CHANGES_REQUESTED** [review]

**Comment by pmconne** [comment]
Do you have local changes for the tests you say are fixed? I don't see relevant commits.

File: full-stack-tests/core/src/frontend/standalone/BackgroundMap.test.ts (hunk 6 above)
Comment (RBBentley) [reply full-stack-tests/core/src/frontend/standalone/BackgroundMap.test.ts]: Yes... removing that was not necessary.  Restored.

**Comment by RBBentley** [comment]
> Do you have local changes for the tests you say are fixed? I don't see relevant commits.

They are there now.

**Comment by RBBentley** [comment]
@pmconne Still reviewing?

**Review by pmconne — APPROVED** [review]

**Comment by pmconne** [comment]
Have these been run through @DStradley's image regression tests? Were any important/problematic differences detected?

**Comment by RBBentley** [comment]
> Have these been run through @DStradley's image regression tests? Were any important/problematic differences detected?

Yes - We've been doing this for last two weeks.   Only significant differences are improved color mixing for thematic on reality models.

**Pull request merged**